### PR TITLE
`GeneralizedDivisiveNormalization` layer

### LINKED ISSUE #85: `GeneralizedDivisiveNormalization` layer
```Python
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""

import functools
from typing import Callable, Optional, Union

import torch
import torch.nn.functional
from torch import Tensor
from torch.nn import Module, Parameter

from ._non_negative_parameterization import NonNegativeParameterization


class GeneralizedDivisiveNormalization(Module):
    """Generalized divisive normalization

    Implements an activation function that is a multivariate generalization of
    the following sigmoid-like function:

        y[i] = x[i] / (beta[i] + sum_j(gamma[j, i] * |x[j]|^alpha))^epsilon

    where ``i`` and ``j`` map over channels. This implementation never sums
    across spatial dimensions. It is similar to local response normalization,
    but much more flexible, as ``alpha``, ``beta``, ``gamma``, and ``epsilon``
    are trainable parameters.

    The method was originally described in:

        | “Density Modeling of Images using a Generalized Normalization
            Transformation”
        | Johannes Ballé, Valero Laparra, Eero P. Simoncelli
        | https://arxiv.org/abs/1511.06281

    and expanded in:

        | “End-to-end Optimized Image Compression”
        | Johannes Ballé, Valero Laparra, Eero P. Simoncelli
        | https://arxiv.org/abs/1611.01704

    Args:
        channels: number of channels in the input.
        inverse: compute the generalized divisive normalization response. If
            ``True``, compute the inverse generalized divisive normalization
            response (one step of fixed point iteration to invert the
            generalized divisive normalization; the division is replaced by
            multiplication).
        rectify: If ``True``, apply a ``ReLU`` non-linearity to the inputs
            before calculating the generalized divisive normalization response.
        alpha_parameter: A ``Tensor`` means that the value of ``alpha`` is
            fixed. A ``Callable`` can be used to determine the value of
            ``alpha`` as a function of some other parameter or tensor. This can
            be a ``Parameter``. ``None`` means that when the layer is
            initialized, a ``NonNegativeParameterization`` layer is created to
            train ``alpha`` (with a minimum value of ``1``). The default is a
            fixed value of ``1``.
        beta_parameter: A ``Tensor`` means that the value of ``beta`` is fixed.
            A ``Callable`` can be used to determine the value of ``beta`` as a
            function of some other parameter or tensor. This can be a
            ``Parameter``. ``None`` means that when the layer is initialized, a
            ``NonNegativeParameterization`` layer is created to train ``beta``
            (with a minimum value of ``1e-6``).
        epsilon_parameter: A ``Tensor`` means that the value of ``epsilon`` is
            fixed. A ``Callable`` can be used to determine the value of
            ``epsilon`` as a function of some other parameter or tensor. This
            can be a ``Parameter``. ``None`` means that when the layer is
            initialized, a ``NonNegativeParameterization`` layer is created to
            train ``epsilon`` (with a minimum value of 1e-6). The default is a
            fixed value of ``1``.
        gamma_parameter: A ``Tensor`` means that the value of ``gamma`` is
            fixed. A ``Callable`` can be used to determine the value of
            ``gamma`` as a function of some other parameter or tensor. This can
            be a ``Parameter``. ``None`` means that when the layer is
            initialized, a ``NonNegativeParameterization`` layer is created to
            train ``gamma``.
        alpha_initializer: initializes the ``alpha`` parameter. Only used if
            ``alpha`` is trained. Defaults to ``1``.
        beta_initializer: initializes the ``beta`` parameter. Only used if
            ``beta`` is created when initializing the layer. Defaults to ``1``.
        epsilon_initializer: initializes the ``epsilon`` parameter. Only used
            if ``epsilon`` is trained. Defaults to ``1``.
        gamma_initializer: initializes the ``gamma`` parameter. Only used if
            ``gamma`` is created when initializing the layer. Defaults to the
            identity multiplied by ``0.1``. A good default value for the
            diagonal is somewhere between ``0`` and ``0.5``. If set to ``0``
            and ``beta`` is initialized as ``1``, the layer is effectively
            initialized to the identity operation.
    """

    def __init__(
        self,
        channels: Union[int, Tensor],
        inverse: bool = False,
        rectify: bool = False,
        alpha_parameter: Union[float, int, Tensor, Parameter] = None,
        beta_parameter: Union[float, int, Tensor, Parameter] = None,
        epsilon_parameter: Union[float, int, Tensor, Parameter] = None,
        gamma_parameter: Union[float, int, Tensor, Parameter] = None,
        alpha_initializer: Optional[Callable[[Tensor], Tensor]] = None,
        beta_initializer: Optional[Callable[[Tensor], Tensor]] = None,
        epsilon_initializer: Optional[Callable[[Tensor], Tensor]] = None,
        gamma_initializer: Optional[Callable[[Tensor], Tensor]] = None,
    ):
        super(GeneralizedDivisiveNormalization, self).__init__()

        self._inverse = inverse

        self._rectify = rectify

        if alpha_parameter is None:
            if alpha_initializer is None:
                alpha_initializer = torch.ones

            self._reparameterized_alpha = NonNegativeParameterization(
                alpha_initializer(channels),
                minimum=1,
            )

            self._alpha_parameter = Parameter(
                self._reparameterized_alpha.initial_value,
            )
        else:
            self._alpha_parameter = alpha_parameter

        if beta_parameter is None:
            if beta_initializer is None:
                beta_initializer = torch.ones

            self._reparameterized_beta = NonNegativeParameterization(
                beta_initializer(channels),
                minimum=1e-6,
            )

            self._beta_parameter = Parameter(
                self._reparameterized_beta.initial_value,
            )
        else:
            self._beta_parameter = beta_parameter

        if epsilon_parameter is None:
            if epsilon_initializer is None:
                epsilon_initializer = torch.ones

            self._reparameterized_epsilon = NonNegativeParameterization(
                epsilon_initializer(channels),
                minimum=1e-6,
            )

            self._epsilon_parameter = Parameter(
                self._reparameterized_epsilon.initial_value,
            )
        else:
            self._epsilon_parameter = epsilon_parameter

        if gamma_parameter is None:
            if gamma_initializer is None:
                gamma_initializer = functools.partial(lambda x: 0.1 * torch.eye(x))

            self._reparameterized_gamma = NonNegativeParameterization(
                gamma_initializer(channels),
                minimum=0,
            )

            self._gamma_parameter = Parameter(
                self._reparameterized_gamma.initial_value,
            )
        else:
            self._gamma_parameter = gamma_parameter

    def forward(self, x: Tensor) -> Tensor:
        _, channels, _, _ = x.size()

        if self._rectify:
            x = torch.nn.functional.relu(x)

        y = torch.nn.functional.conv2d(
            x ** 2,
            torch.reshape(
                self._reparameterized_gamma(self._gamma_parameter),
                (channels, channels, 1, 1),
            ),
            self._reparameterized_beta(self._beta_parameter),
        )

        if self._inverse:
            return x * torch.sqrt(y)

        return x * torch.rsqrt(y)
```

```Python
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""

import torch
import torch.testing

from neuralcompression.layers import GeneralizedDivisiveNormalization


class TestGeneralizedDivisiveNormalization:
    def test_backward(self):
        x = torch.rand((1, 32, 16, 16), requires_grad=True)

        normalization = GeneralizedDivisiveNormalization(32)

        normalized = normalization(x)

        normalized.backward(x)

        assert normalized.shape == x.shape

        assert x.grad is not None

        assert x.grad.shape == x.shape

        torch.testing.assert_allclose(
            x / torch.sqrt(1 + 0.1 * (x ** 2)),
            normalized,
        )

        normalization = GeneralizedDivisiveNormalization(
            32,
            inverse=True,
        )

        normalized = normalization(x)

        normalized.backward(x)

        assert normalized.shape == x.shape

        assert x.grad is not None

        assert x.grad.shape == x.shape

        torch.testing.assert_allclose(
            x * torch.sqrt(1 + 0.1 * (x ** 2)),
            normalized,
        )
```

Prerequisites

- [x] #86

### PR DESCRIPTION
**Pull request by 0x00b1**
closes #85 

Prerequisites

- [x] #96

### CODE REVIEW

**Comment by mmuckley** [comment]
@0x00b1 could you rebase to #96 or rebase after merge?

**Comment by 0x00b1** [comment]
@mmuckley rebased!

File: neuralcompression/layers/_generalized_divisive_normalization.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            else:
+                gamma_parameter = torch.tensor(gamma_parameter)
+
+                self.gamma = Parameter(gamma_parameter)
+
+    def forward(self, x: Tensor) -> Tensor:
+        _, channels, _, _ = x.size()
+
+        if self._rectify:
+            x = torch.nn.functional.relu(x)
```
Comment (mmuckley) [inline neuralcompression/layers/_generalized_divisive_normalization.py:217]: We don't need this here - we should just make users use their own `torch.nn.ReLU()` layer.

File: neuralcompression/layers/_generalized_divisive_normalization.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        beta_initializer: initializes the ``beta`` parameter. Only used if
+            ``beta`` is created when initializing the layer. Defaults to ``1``.
+        epsilon_initializer: initializes the ``epsilon`` parameter. Only used
+            if ``epsilon`` is trained. Defaults to ``1``.
+        gamma_initializer: initializes the ``gamma`` parameter. Only used if
+            ``gamma`` is created when initializing the layer. Defaults to the
+            identity multiplied by ``0.1``. A good default value for the
+            diagonal is somewhere between ``0`` and ``0.5``. If set to ``0``
+            and ``beta`` is initialized as ``1``, the layer is effectively
+            initialized to the identity operation.
```
Comment (mmuckley) [inline neuralcompression/layers/_generalized_divisive_normalization.py:91]: This API is too complicated. One thing in particular I would like to change is divergent behavior based on input type. The best option is for every parameter of the class to have one type. The second best option is to have union types where the types in the Union are functionally similar (think `Path` and `os.path_like`). A `Tensor` and a `Callable` are completely different though.

The reason it is too complicated is there are too many features. If you want all of these features, I would split them out into helper functions and document the spectrum of helper functions available to the user. Another way to reduce the API complication is to use a `NamedTuple` that clearly semantically spaces out what you are putting in as the parameters of the class.

**Review by mmuckley — CHANGES_REQUESTED** [review]
I think the API is too complicated. I would split out the functionality that you're going for into a set of helper functions and try to guide the user to the helper functions via documentation. In particular I don't think it's a good idea to have divergent behavior based on input type. This is a pattern that can lead to feature bloat and unwieldly code in the long run. Just make a new class or a new function that does what you're going for.

File: neuralcompression/layers/_generalized_divisive_normalization.py (hunk 1 above)
Comment (0x00b1) [reply neuralcompression/layers/_generalized_divisive_normalization.py:217]: Good point.

Comment (0x00b1) [reply neuralcompression/layers/_generalized_divisive_normalization.py]: Removed!

**Review by mmuckley — APPROVED** [review]
Overall looks good. For complicated APIs like this it might be good to include some `Examples` in the docstring to illustrate standard use cases.

I still think we could iterate on this in the future, but it solves my earlier issue of type divergence.

**Pull request merged**